File: types/backtest.ts
/**
 * Mirrors the backend's Backtesting Engine response shapes (see
 * backend/app/schemas/backtest.py). The engine replays the unmodified
 * Sprint 3 AI Decision Engine bar by bar over historical candles — no
 * look-ahead, no simulated trade execution against Binance.
 */
export type TradeDirection = "LONG" | "SHORT";
export type ExitReason = "TP1" | "SL" | "END_OF_DATA";
export type BacktestRunStatus = "PENDING" | "RUNNING" | "COMPLETED" | "FAILED";

export const BACKTEST_PERIOD_DAYS = [30, 90, 180, 365] as const;
export type BacktestPeriodDays = (typeof BACKTEST_PERIOD_DAYS)[number];

export interface BacktestRunRequest {
  symbol: string;
  timeframe: string;
  periodDays: BacktestPeriodDays;
  initialBalance?: number;
  commissionBps?: number;
  slippageBps?: number;
  riskPerTradePercent?: number;
}

export interface PerformanceMetrics {
  totalTrades: number;
  winningTrades: number;
  losingTrades: number;
  totalReturnPercent: number;
  netProfit: number;
  grossProfit: number;
  grossLoss: number;
  winRate: number;
  lossRate: number;
  avgWin: number;
  avgLoss: number;
  profitFactor: number;
  expectancy: number;
  avgTradeDurationSeconds: number;
}

export interface RiskMetrics {
  maxDrawdownPercent: number;
  maxDrawdownDurationSeconds: number;
  recoveryFactor: number;
  sharpeRatio: number;
  sortinoRatio: number;
  calmarRatio: number;
  avgRiskReward: number;
  finalBalance: number;
  peakBalance: number;
}

export interface BacktestTrade {
  id: string;
  symbol: string;
  direction: TradeDirection;
  entryTime: number;
  entryPrice: number;
  exitTime: number;
  exitPrice: number;
  quantity: number;
  pnl: number;
  pnlPercent: number;
  exitReason: ExitReason;
  durationSeconds: number;
  decisionScore: number;
  confidence: number;
  plannedRiskReward: number;
}

export interface EquityPoint {
  time: number;
  balance: number;
  drawdownPercent: number;
  cumulativePnl: number;
  tradeCount: number;
}

export interface BacktestRun {
  id: string;
  symbol: string;
  timeframe: string;
  periodDays: number;
  startTime: number;
  endTime: number;
  status: BacktestRunStatus;
  strategyVersionName: string;
  initialBalance: number;
  commissionBps: number;
  slippageBps: number;
  errorMessage: string | null;
  startedAt: number | null;
  completedAt: number | null;
  durationMs: number | null;
}

export interface BacktestMetrics {
  performance: PerformanceMetrics;
  risk: RiskMetrics;
}

export interface BacktestReport {
  run: BacktestRun;
  performance: PerformanceMetrics;
  risk: RiskMetrics;
}

File: backend/app/schemas/backtest.py
from typing import Literal

from app.schemas.base import CamelModel

TradeDirection = Literal["LONG", "SHORT"]
ExitReason = Literal["TP1", "SL", "END_OF_DATA"]
RunStatus = Literal["PENDING", "RUNNING", "COMPLETED", "FAILED"]


class BacktestRunRequest(CamelModel):
    symbol: str
    timeframe: str
    period_days: int
    initial_balance: float = 10_000.0
    commission_bps: float = 4.0
    slippage_bps: float = 2.0
    risk_per_trade_percent: float = 1.0


class PerformanceMetricsOut(CamelModel):
    total_trades: int
    winning_trades: int
    losing_trades: int
    total_return_percent: float
    net_profit: float
    gross_profit: float
    gross_loss: float
    win_rate: float
    loss_rate: float
    avg_win: float
    avg_loss: float
    profit_factor: float
    expectancy: float
    avg_trade_duration_seconds: float


class RiskMetricsOut(CamelModel):
    max_drawdown_percent: float
    max_drawdown_duration_seconds: int
    recovery_factor: float
    sharpe_ratio: float
    sortino_ratio: float
    calmar_ratio: float
    avg_risk_reward: float
    final_balance: float
    peak_balance: float


class BacktestTradeOut(CamelModel):
    id: str
    symbol: str
    direction: TradeDirection
    entry_time: int
    entry_price: float
    exit_time: int
    exit_price: float
    quantity: float
    pnl: float
    pnl_percent: float
    exit_reason: ExitReason
    duration_seconds: int
    decision_score: float
    confidence: float
    planned_risk_reward: float


class EquityPointOut(CamelModel):
    time: int
    balance: float
    drawdown_percent: float
    cumulative_pnl: float
    trade_count: int


class BacktestRunOut(CamelModel):
    id: str
    symbol: str
    timeframe: str
    period_days: int
    start_time: int
    end_time: int
    status: RunStatus
    strategy_version_name: str
    initial_balance: float
    commission_bps: float
    slippage_bps: float
    error_message: str | None
    started_at: int | None
    completed_at: int | None
    duration_ms: int | None


class BacktestMetricsOut(CamelModel):
    """Just the numbers — what `/metrics/{run_id}` returns. `/report`
    returns this plus the run's own metadata (see `BacktestReportOut`)."""

    performance: PerformanceMetricsOut
    risk: RiskMetricsOut


class BacktestReportOut(CamelModel):
    """Everything about one run in a single payload — what `/report`
    returns, and what `/run` returns immediately after a run completes so
    the frontend doesn't need a second round-trip to show results."""

    run: BacktestRunOut
    performance: PerformanceMetricsOut
    risk: RiskMetricsOut
